routers/api/v1/repo/tree.go

```diff
@@ -87,30 +106,47 @@ func GetTreeBySHA(ctx *context.APIContext, sha string) *gitea.GitTreeResponse {
 	// 40 is the size of the sha1 hash in hexadecimal format.
 	copyPos := len(treeURL) - 40
 
-	if len(entries) > 1000 {
-		tree.Entries = make([]gitea.GitEntry, 1000)
+	page := ctx.QueryInt("page")
+	perPage := ctx.QueryInt("per_page")
+	if perPage <= 0 || perPage > setting.API.DefaultGitTreesPerPage {
+		perPage = setting.API.DefaultGitTreesPerPage
+	}
+	if page <= 0 {
+		page = 1
+	}
+	tree.Page = page
+	tree.TotalCount = len(entries)
+	rangeStart := perPage * (page - 1)
+	if rangeStart >= len(entries) {
+		return tree
+	}
+	var rangeEnd int
+	if len(entries) > perPage {
+		tree.Truncated = true
+	}
+	if rangeStart+perPage < len(entries) {
+		rangeEnd = rangeStart + perPage
 	} else {
-		tree.Entries = make([]gitea.GitEntry, len(entries))
+		rangeEnd = len(entries)
 	}
-	for e := range entries {
-		if e > 1000 {
-			tree.Truncated = true
-			break
-		}
-
-		tree.Entries[e].Path = entries[e].Name()
-		tree.Entries[e].Mode = fmt.Sprintf("%06x", entries[e].Mode())
-		tree.Entries[e].Type = string(entries[e].Type)
-		tree.Entries[e].Size = entries[e].Size()
-		tree.Entries[e].SHA = entries[e].ID.String()
+	tree.Entries = make([]gitea.GitEntry, rangeEnd-rangeStart)
+	for e := rangeStart; e < rangeEnd; e++ {
+		i := e - rangeStart
+		tree.Entries[i].Path = entries[e].Name()
+		tree.Entries[i].Mode = fmt.Sprintf("%06x", entries[e].Mode())
+		tree.Entries[i].Type = string(entries[e].Type)
+		tree.Entries[i].Size = entries[e].Size()
+		tree.Entries[i].SHA = entries[e].ID.String()
 
 		if entries[e].IsDir() {
 			copy(treeURL[copyPos:], entries[e].ID.String())
-			tree.Entries[e].URL = string(treeURL[:])
+			tree.Entries[i].URL = string(treeURL[:])
 		} else {
 			copy(blobURL[copyPos:], entries[e].ID.String())
-			tree.Entries[e].URL = string(blobURL[:])
+			tree.Entries[i].URL = string(blobURL[:])
 		}
 	}
+	fmt.Printf("HEREEEEE:\n")
```

Comment:
Assuming these will be removed before merge, just mentioning it just in case they get forgotten

Reply:
Yeah, currently working on it...you guys are fast.

Reply:
Was working on trying to write my first test and was just going to see if it worked even though past locally. Now I have removed those print statements.